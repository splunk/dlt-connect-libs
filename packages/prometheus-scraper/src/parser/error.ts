export class ParserError extends Error {
    public readonly cause?: Error;
    constructor(message: string, cause?: Error) {
        super(message);
        this.name = this.constructor.name;
        this.message = message;
        if (cause) {
            this.stack = cause.stack;
            this.cause = cause;
        } else if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, this.constructor);
        } else {
            this.stack = new Error(message).stack;
        }
    }
}
