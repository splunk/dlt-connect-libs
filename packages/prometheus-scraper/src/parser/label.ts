import { ParserError } from './error';

export type Label = {
    name: string;
    value: string;
};

export function parseLabelString(labelString: string): Label[] {
    if (labelString === '') {
        return [];
    }
    const result: Label[] = [];

    if (!(labelString.startsWith('{') && labelString.endsWith('}'))) {
        throw new ParserError('Invalid label string');
    }
    let cur = labelString.slice(1, -1);

    while (cur.length) {
        const eqIdx = cur.indexOf('=');
        if (eqIdx < 0) {
            throw new ParserError('Invalid label string');
        }
        const label = cur.slice(0, eqIdx);
        if (!/^[a-zA-Z_][a-zA-Z0-9_]*$/.test(label)) {
            throw new ParserError(`Invalid label ${JSON.stringify(label)}`);
        }

        let value = '';
        let valueIdx = eqIdx + 1;
        let escaped = false;

        if (cur[valueIdx] !== '"') {
            throw new ParserError('Invalid label value (not quoted)');
        }

        valueIdx++;
        for (; ; valueIdx++) {
            if (valueIdx > cur.length) {
                throw new ParserError('Premature end of label string');
            }
            const ch = cur[valueIdx];
            if (escaped) {
                escaped = false;
                if (ch === '"') {
                    value += '"';
                } else if (ch === 'n') {
                    value += '\n';
                } else if (ch === '\\') {
                    value += '\\';
                } else {
                    throw new ParserError(`Invalid escape sequence: "\\${ch}"`);
                }
                continue;
            }
            if (ch === '"') {
                break;
            }
            if (ch === '\\') {
                escaped = true;
                continue;
            }
            value += ch;
        }

        result.push({ name: label, value });
        cur = cur.slice(valueIdx + 1);

        if (cur[0] === ',') {
            cur = cur.slice(1);
        } else if (cur.length) {
            throw new ParserError('Invalid label string, missing comma after value');
        }
    }

    return result;
}
