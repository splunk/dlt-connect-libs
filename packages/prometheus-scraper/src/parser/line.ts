import { createModuleDebug } from '@splunkdlt/debug-logging';
import { MetricName } from '../types';
import { ParserError } from './error';
import { Label, parseLabelString } from './label';

const { debug } = createModuleDebug('@splunkdlt/prometheus-scraper:parser:line');

type MetricType = 'counter' | 'histogram' | 'gauge' | 'summary' | 'untyped';

type Line =
    | { type: 'empty' }
    | { type: 'comment'; commentText: string }
    | { type: 'help'; metricName: MetricName; helpText: string }
    | { type: 'type-info'; metricName: MetricName; typeName: MetricType }
    | { type: 'metric'; metricName: MetricName; labels: Label[]; value: string; timestamp: string };

export function parseLine(line: string): Line {
    if (!line?.trim()) {
        return { type: 'empty' };
    }

    const trimmed = line.trim();

    if (trimmed.startsWith('#')) {
        let m = trimmed.match(/^#\s*HELP\s+([a-zA-Z_:][a-zA-Z0-9_:]*)\s+(.+)/);
        if (m != null) {
            return { type: 'help', metricName: m[1], helpText: m[2] };
        }

        m = trimmed.match(/^#\s*TYPE\s+([a-zA-Z_:][a-zA-Z0-9_:]*)\s+(counter|histogram|gauge|summary)/);
        if (m != null) {
            return { type: 'type-info', metricName: m[1], typeName: m[2] as MetricType };
        }

        return {
            type: 'comment',
            commentText: trimmed.slice(1).trimStart(),
        };
    }

    const m = trimmed.match(/^([a-zA-Z_:][a-zA-Z0-9_:]*)( ?\{.+?\}|)\s+([\d\.e+-]+|[+-][iI]nf|Na[Nn])(?:\s+(-?\d+))?$/);
    if (m != null) {
        return {
            type: 'metric',
            metricName: m[1],
            labels: parseLabelString(m[2].trim()),
            value: m[3],
            timestamp: m[4],
        };
    }

    debug('Encountered invalid line: %o', line);
    throw new ParserError(`Invalid line: ${JSON.stringify(line)}`);
}
