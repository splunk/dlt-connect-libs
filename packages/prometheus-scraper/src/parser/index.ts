/*
 * Parser for Prometheus text format
 * https://github.com/Showmax/prometheus-docs/blob/master/content/docs/instrumenting/exposition_formats.md
 */

import { createModuleDebug } from '@splunkdlt/debug-logging';
import { BucketedMetric, Metric, MetricValue } from '../types';
import { ParserError } from './error';
import { Label } from './label';
import { parseLine } from './line';

const { warn } = createModuleDebug('@splunkdlt/prometheus-scraper:parser');

export interface ParserOptions {
    failOnInvalidLine: boolean;
    failOnDuplicateTypeInfo: boolean;
    failOnDuplicateHelp: boolean;
}

export const PARSER_DEFAULTS: ParserOptions = {
    failOnInvalidLine: true,
    failOnDuplicateHelp: false,
    failOnDuplicateTypeInfo: false,
};

type EmitFn = (metric: Metric | null) => void;

class StatefulParser {
    constructor(private emit: EmitFn, private options: ParserOptions) {}
    private readonly typeMap = new Map<string, string>();
    private readonly helpMap = new Map<string, string>();
    private currentMultiSample: BucketedMetric | null = null;
    private cachedLabelMap: { labels: Label[]; result: Map<string, string> } | null = null;

    flushCurrentMultiSample = () => {
        if (this.currentMultiSample != null) {
            this.emit(this.currentMultiSample);
            this.currentMultiSample = null;
        }
    };

    labelMap(labels: Label[]): Map<string, string> {
        if (labels === this.cachedLabelMap?.labels) {
            return this.cachedLabelMap.result;
        }
        const result = new Map<string, string>();
        for (const { name, value } of labels) {
            result.set(name, value);
        }
        this.cachedLabelMap = {
            labels,
            result,
        };
        return result;
    }

    labelMatch(a: Label[], b: Label[], exclude: 'le' | 'quantile'): boolean {
        const af = a.filter((l) => l.name !== exclude);
        if (af.length !== b.length) {
            return false;
        }
        const bm = this.labelMap(b);
        return af.every(({ name, value }) => bm.get(name) === value);
    }

    processLine = (lineStr: string) => {
        const line = parseLine(lineStr);

        switch (line.type) {
            case 'comment':
            case 'empty':
                return;

            case 'type-info':
                if (this.options.failOnDuplicateTypeInfo && this.typeMap.has(line.metricName)) {
                    throw new Error(`Duplicate type declaraction for metric ${JSON.stringify(line.metricName)}`);
                }
                this.typeMap.set(line.metricName, line.typeName);
                break;

            case 'help':
                if (this.options.failOnDuplicateHelp && this.helpMap.has(line.metricName)) {
                    throw new Error(`Duplicate help declaraction for metric ${JSON.stringify(line.metricName)}`);
                }
                this.helpMap.set(line.metricName, line.helpText);
                break;

            case 'metric':
                //
                let name = line.metricName;
                let type = this.typeMap.get(name);
                if (type == null) {
                    if (name.endsWith('_bucket')) {
                        const t = this.typeMap.get(name.slice(0, -7));
                        if (t === 'histogram') {
                            name = name.slice(0, -7);
                            type = t;
                        }
                    }
                    if (name.endsWith('_sum')) {
                        const t = this.typeMap.get(name.slice(0, -4));
                        if (t === 'histogram' || t === 'summary') {
                            name = name.slice(0, -4);
                            type = t;
                        }
                    }
                    if (name.endsWith('_count')) {
                        const t = this.typeMap.get(name.slice(0, -6));
                        if (t === 'histogram' || t === 'summary') {
                            name = name.slice(0, -6);
                            type = t;
                        }
                    }
                }

                switch (type) {
                    case 'counter':
                    case 'untyped':
                    case 'gauge':
                        this.flushCurrentMultiSample();
                        this.emit({
                            name,
                            type,
                            labels: line.labels,
                            value: line.value,
                            help: this.helpMap.get(name),
                            timestamp: line.timestamp,
                        });
                        break;
                    case 'histogram':
                        this.processBucketedMetric(name, line, 'histogram', 'le', '_bucket');
                        break;
                    case 'summary':
                        this.processBucketedMetric(name, line, 'summary', 'quantile', '');
                        break;
                }

                break;

            default:
                throw new ParserError(`Received invalid line of type ${type}`);
        }
    };

    processBucketedMetric = (
        name: string,
        line: { metricName: string; type: string; labels: Label[]; value: MetricValue },
        type: 'histogram' | 'summary',
        bucketLabel: 'le' | 'quantile',
        bucketSuffix: '' | '_bucket'
    ) => {
        if (
            this.currentMultiSample != null &&
            (name !== this.currentMultiSample.name ||
                !this.labelMatch(line.labels, this.currentMultiSample.labels, bucketLabel))
        ) {
            this.flushCurrentMultiSample();
        }
        if (this.currentMultiSample == null) {
            this.currentMultiSample = {
                name,
                type,
                labels: line.labels.filter((label) => label.name !== bucketLabel),
                buckets: [],
                sum: 'NaN',
                count: 'NaN',
            };
        }
        const cur = this.currentMultiSample;
        const suffix = line.metricName.slice(name.length);
        switch (suffix) {
            case bucketSuffix:
                const threshold = line.labels.find((l) => l.name === bucketLabel)?.value;
                if (threshold == null) {
                    warn('Encountered bucketed metric without threshold label "%s"', bucketLabel);
                    break;
                }
                cur.buckets.push([threshold, line.value]);
                break;
            case '_count':
                cur.count = line.value;
                break;
            case '_sum':
                cur.sum = line.value;
                break;
            default:
                throw new ParserError(`Unexpected metric name suffix: ${JSON.stringify(suffix)}`);
        }
    };

    done = () => {
        this.flushCurrentMultiSample();
        this.emit(null);
    };
}

export function parseText(text: string, options: ParserOptions = PARSER_DEFAULTS): Metric[] {
    const result: Metric[] = [];
    const parser = new StatefulParser((metric) => {
        if (metric != null) result.push(metric);
    }, options);
    try {
        text.split('\n').forEach(parser.processLine);
        parser.done();
    } catch (e) {
        if (e instanceof ParserError) {
            if (options.failOnInvalidLine) {
                throw new ParserError(`Error parsing Prometheus metrics: ${e.message}`);
            } else {
                warn('Ignoring invalid line', e);
            }
        } else {
            throw new ParserError(`Unexpected error parsing Prometheus metrics: ${e}`, e);
        }
    }
    return result;
}
