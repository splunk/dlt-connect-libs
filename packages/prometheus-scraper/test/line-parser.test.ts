import { readFileSync } from 'fs';
import { join } from 'path';
import { parseLine } from '../src/parser/line';

describe('parseLine', () => {
    it('parses empty line', () => {
        expect(parseLine('')).toEqual({ type: 'empty' });
    });

    it('parses HELP line', () => {
        expect(parseLine('# HELP some_metric_name This is some help text')).toEqual({
            type: 'help',
            metricName: 'some_metric_name',
            helpText: 'This is some help text',
        });
    });

    it('parses TYPE line', () => {
        expect(parseLine('# TYPE some_metric_name counter')).toEqual({
            type: 'type-info',
            metricName: 'some_metric_name',
            typeName: 'counter',
        });
    });

    it('parses metric line', () => {
        expect(parseLine('some_metric_name 123')).toEqual({
            type: 'metric',
            metricName: 'some_metric_name',
            value: '123',
            labels: [],
        });
        expect(parseLine('some_metric_name{foo="bar"} 123')).toEqual({
            type: 'metric',
            metricName: 'some_metric_name',
            value: '123',
            labels: [{ name: 'foo', value: 'bar' }],
        });
        expect(parseLine('some_metric_name{foo="bar"} 123 45678')).toEqual({
            type: 'metric',
            metricName: 'some_metric_name',
            value: '123',
            labels: [{ name: 'foo', value: 'bar' }],
            timestamp: '45678',
        });
    });

    it('throws error on invalid line', () => {
        expect(() => parseLine('THIS IS SPARTA!')).toThrowErrorMatchingInlineSnapshot(
            `"Invalid line: \\"THIS IS SPARTA!\\""`
        );
    });

    it('parses fixture files', async () => {
        const contents = readFileSync(join(__dirname, 'fixtures/example1.txt'), { encoding: 'utf-8' });
        expect(contents.split('\n').map(parseLine)).toMatchSnapshot();
    });
});
