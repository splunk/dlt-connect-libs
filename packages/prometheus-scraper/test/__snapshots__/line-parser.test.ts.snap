// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseLine parses fixture files 1`] = `
Array [
  Object {
    "helpText": "The total number of HTTP requests.",
    "metricName": "http_requests_total",
    "type": "help",
  },
  Object {
    "metricName": "http_requests_total",
    "type": "type-info",
    "typeName": "counter",
  },
  Object {
    "labels": Array [
      Object {
        "name": "method",
        "value": "post",
      },
      Object {
        "name": "code",
        "value": "200",
      },
    ],
    "metricName": "http_requests_total",
    "timestamp": "1395066363000",
    "type": "metric",
    "value": "1027",
  },
  Object {
    "labels": Array [
      Object {
        "name": "method",
        "value": "post",
      },
      Object {
        "name": "code",
        "value": "400",
      },
    ],
    "metricName": "http_requests_total",
    "timestamp": "1395066363000",
    "type": "metric",
    "value": "3",
  },
  Object {
    "type": "empty",
  },
  Object {
    "commentText": "Escaping in label values:",
    "type": "comment",
  },
  Object {
    "labels": Array [
      Object {
        "name": "path",
        "value": "C:\\\\DIR\\\\FILE.TXT",
      },
      Object {
        "name": "error",
        "value": "Cannot find file:
\\"FILE.TXT\\"",
      },
    ],
    "metricName": "msdos_file_access_time_seconds",
    "timestamp": undefined,
    "type": "metric",
    "value": "1.458255915e9",
  },
  Object {
    "type": "empty",
  },
  Object {
    "commentText": "Minimalistic line:",
    "type": "comment",
  },
  Object {
    "labels": Array [],
    "metricName": "metric_without_timestamp_and_labels",
    "timestamp": undefined,
    "type": "metric",
    "value": "12.47",
  },
  Object {
    "type": "empty",
  },
  Object {
    "commentText": "A weird metric from before the epoch:",
    "type": "comment",
  },
  Object {
    "labels": Array [
      Object {
        "name": "problem",
        "value": "division by zero",
      },
    ],
    "metricName": "something_weird",
    "timestamp": "-3982045",
    "type": "metric",
    "value": "+Inf",
  },
  Object {
    "type": "empty",
  },
  Object {
    "commentText": "A histogram, which has a pretty complex representation in the text format:",
    "type": "comment",
  },
  Object {
    "helpText": "A histogram of the request duration.",
    "metricName": "http_request_duration_seconds",
    "type": "help",
  },
  Object {
    "metricName": "http_request_duration_seconds",
    "type": "type-info",
    "typeName": "histogram",
  },
  Object {
    "labels": Array [
      Object {
        "name": "le",
        "value": "0.05",
      },
    ],
    "metricName": "http_request_duration_seconds_bucket",
    "timestamp": undefined,
    "type": "metric",
    "value": "24054",
  },
  Object {
    "labels": Array [
      Object {
        "name": "le",
        "value": "0.1",
      },
    ],
    "metricName": "http_request_duration_seconds_bucket",
    "timestamp": undefined,
    "type": "metric",
    "value": "33444",
  },
  Object {
    "labels": Array [
      Object {
        "name": "le",
        "value": "0.2",
      },
    ],
    "metricName": "http_request_duration_seconds_bucket",
    "timestamp": undefined,
    "type": "metric",
    "value": "100392",
  },
  Object {
    "labels": Array [
      Object {
        "name": "le",
        "value": "0.5",
      },
    ],
    "metricName": "http_request_duration_seconds_bucket",
    "timestamp": undefined,
    "type": "metric",
    "value": "129389",
  },
  Object {
    "labels": Array [
      Object {
        "name": "le",
        "value": "1",
      },
    ],
    "metricName": "http_request_duration_seconds_bucket",
    "timestamp": undefined,
    "type": "metric",
    "value": "133988",
  },
  Object {
    "labels": Array [
      Object {
        "name": "le",
        "value": "+Inf",
      },
    ],
    "metricName": "http_request_duration_seconds_bucket",
    "timestamp": undefined,
    "type": "metric",
    "value": "144320",
  },
  Object {
    "labels": Array [],
    "metricName": "http_request_duration_seconds_sum",
    "timestamp": undefined,
    "type": "metric",
    "value": "53423",
  },
  Object {
    "labels": Array [],
    "metricName": "http_request_duration_seconds_count",
    "timestamp": undefined,
    "type": "metric",
    "value": "144320",
  },
  Object {
    "type": "empty",
  },
  Object {
    "commentText": "Finally a summary, which has a complex representation, too:",
    "type": "comment",
  },
  Object {
    "helpText": "A summary of the response latency.",
    "metricName": "telemetry_requests_metrics_latency_microseconds",
    "type": "help",
  },
  Object {
    "metricName": "telemetry_requests_metrics_latency_microseconds",
    "type": "type-info",
    "typeName": "summary",
  },
  Object {
    "labels": Array [
      Object {
        "name": "quantile",
        "value": "0.01",
      },
    ],
    "metricName": "telemetry_requests_metrics_latency_microseconds",
    "timestamp": undefined,
    "type": "metric",
    "value": "3102",
  },
  Object {
    "labels": Array [
      Object {
        "name": "quantile",
        "value": "0.05",
      },
    ],
    "metricName": "telemetry_requests_metrics_latency_microseconds",
    "timestamp": undefined,
    "type": "metric",
    "value": "3272",
  },
  Object {
    "labels": Array [
      Object {
        "name": "quantile",
        "value": "0.5",
      },
    ],
    "metricName": "telemetry_requests_metrics_latency_microseconds",
    "timestamp": undefined,
    "type": "metric",
    "value": "4773",
  },
  Object {
    "labels": Array [
      Object {
        "name": "quantile",
        "value": "0.9",
      },
    ],
    "metricName": "telemetry_requests_metrics_latency_microseconds",
    "timestamp": undefined,
    "type": "metric",
    "value": "9001",
  },
  Object {
    "labels": Array [
      Object {
        "name": "quantile",
        "value": "0.99",
      },
    ],
    "metricName": "telemetry_requests_metrics_latency_microseconds",
    "timestamp": undefined,
    "type": "metric",
    "value": "76656",
  },
  Object {
    "labels": Array [],
    "metricName": "telemetry_requests_metrics_latency_microseconds_sum",
    "timestamp": undefined,
    "type": "metric",
    "value": "1.7560473e+07",
  },
  Object {
    "labels": Array [],
    "metricName": "telemetry_requests_metrics_latency_microseconds_count",
    "timestamp": undefined,
    "type": "metric",
    "value": "2693",
  },
  Object {
    "type": "empty",
  },
]
`;
